traits/observers/_list_item_observer.py

```diff
@@ -0,0 +1,219 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+from traits.observers._list_change_event import list_event_factory
+from traits.observers._i_observer import IObserver
+from traits.observers._trait_event_notifier import TraitEventNotifier
+from traits.observers._observe import add_or_remove_notifiers
+from traits.observers._observer_change_notifier import ObserverChangeNotifier
+from traits.trait_list_object import TraitList
+
+
+@IObserver.register
+class ListItemObserver:
+    """ Observer for observing mutations on a list.
+
+    Parameters
+    ----------
+    notify : boolean
+        Whether to notify for changes.
+    optional : boolean
+        If False, this observer will complain if the incoming object is not
+        an observable list. If True and the incoming object is not a list, this
+        observer will do nothing. Useful for the 'items' keyword in the text
+        parser, where the source container type is ambiguous.
+    """
+
+    def __init__(self, *, notify, optional):
+        self._notify = notify
+        self._optional = optional
+
+    def __hash__(self):
+        """ Return a hash of this object."""
+        return hash((type(self), self.notify))
```

Comment:
Well spotted. That would be a nasty bug. Thank you!

Reply:
Actually, it would not be that too bad because equal hash does not mean equal objects. It is for this reason I could not write a failing test for this.